src/QsCompiler/QirGeneration/ScopeManager.cs

```diff
@@ -126,28 +126,106 @@ private static IEnumerable<(IValue, bool)> Expand(Func<ITypeRef, string?> getFun
+                    this.decreaseCounts(ReferenceCountUpdateFunctionForType, change);
+                }
+            }
+
+            /// <summary>
+            /// Increases the alias count of the assigned value immediately.
+            /// Increases the reference count immediately if the assigned value is accessed via a local identifier,
+            /// and otherwise removes a pending reference count decrease at the end of the scope if possible.
+            /// If <paramref name="isFromLocalId"/> is false and no reference count decrease is queued,
+            /// increases the reference count immediately.
+            /// </summary>
+            internal void AssignToMutable(bool isFromLocalId, IValue value, bool shallow = false)
+            {
+                var change = new[] { (value, !shallow) };
+                this.increaseCounts(AliasCountUpdateFunctionForType, change);
+
+                // If the assigned value is accessed via a local identifier (either directly or e.g. in the form
+                // of an item access expression), then we need to make sure to apply the reference count increase
+                // immediately to avoid that the bound value may be deallocated if either that local identifier
+                // or the mutable variable is rebound.
+                if (isFromLocalId || !this.TryRemoveValue(value, !shallow))
+                {
+                    // The corresponding decrease is automatically done for values assigned to mutable variables
+                    // when they go out of scope or are rebound.
+                    this.increaseCounts(ReferenceCountUpdateFunctionForType, change);
+                }
+            }
+
             /// <summary>
             /// Registers the given variable name with the scope.
             /// Increases the alias and reference count for the value if necessary,
-            /// and ensures that both are decreased again when the variable goes out of scope.
+            /// depending on whether the value is accessed via a mutable variable or assigned to a mutable variable.
+            /// Ensures that both are decreased again when the variable goes out of scope.
             /// The registered variable is mutable if the passed value is a pointer value.
             /// </summary>
-            public void RegisterVariable(string varName, IValue value)
+            public void RegisterVariable(string varName, IValue value, IValue? fromLocalId)
             {
                 this.variables.Add(varName, value);
 
-                // Since the value is necessarily created in the current or a parent scope,
-                // it won't go out of scope before the variable does.
-                // There is hence no need to increase the reference count if the variable is never rebound.
                 var change = new[] { (value, true) }; // true refers to whether the change also applies to inner items
```

Comment:
The comment above refers to whether or not it is a shallow assignment?

Reply:
No, good point - this applies to the entirety of the logic. I'll make it more apparent.